crates/bevy_reflect/src/path/access.rs

```diff
@@ -0,0 +1,240 @@
+use std::fmt;
+
+use super::ReflectPathError;
+use crate::{Reflect, ReflectMut, ReflectRef, VariantType};
+use thiserror::Error;
+
+type InnerResult<T> = Result<Option<T>, AccessError<'static>>;
+
+#[derive(Debug, PartialEq, Eq, Error)]
+pub enum AccessError<'a> {
+    #[error(
+        "the current {ty} doesn't have the {} {}",
+        access.kind(),
+        access.display_value(),
+    )]
+    Access { ty: Type, access: AccessRef<'a> },
+
+    #[error("invalid type shape: expected {expected} but found a reflect {actual}")]
+    Type { expected: Type, actual: Type },
+
+    #[error("invalid enum access: expected {expected} variant but found {actual} variant")]
+    Enum { expected: Type, actual: Type },
+}
+
+impl<'a> AccessError<'a> {
+    fn with_offset(self, offset: usize) -> ReflectPathError<'a> {
+        let error = self;
+        ReflectPathError::InvalidAccess { offset, error }
+    }
+}
+impl AccessError<'static> {
+    fn bad_enum(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        AccessError::Enum { expected, actual }
+    }
+    fn bad_type(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        AccessError::Type { expected, actual }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Type {
```

Comment:
Doc strings please: linking to the types / traits for each enum variant would be super useful.